datafusion-examples/examples/file_stream_provider.rs

```diff
@@ -0,0 +1,185 @@
+        .with_header(true);
+    let config = StreamConfig::new(Arc::new(source)).with_order(sort);
+    Arc::new(StreamTable::new(Arc::new(config)))
+}
+
+fn create_fifo_file(tmp_dir: &TempDir, file_name: &str) -> Result<PathBuf> {
+    let file_path = tmp_dir.path().join(file_name);
+    // Simulate an infinite environment via a FIFO file
+    if let Err(e) = unistd::mkfifo(&file_path, stat::Mode::S_IRWXU) {
+        exec_err!("{}", e)
+    } else {
+        Ok(file_path)
+    }
+}
+
+fn write_to_fifo(
+    mut file: &File,
+    line: &str,
+    ref_time: Instant,
+    broken_pipe_timeout: Duration,
+) -> Result<()> {
+    // We need to handle broken pipe error until the reader is ready. This
+    // is why we use a timeout to limit the wait duration for the reader.
+    // If the error is different than broken pipe, we fail immediately.
+    while let Err(e) = file.write_all(line.as_bytes()) {
+        if e.raw_os_error().unwrap() == 32 {
+            let interval = Instant::now().duration_since(ref_time);
+            if interval < broken_pipe_timeout {
+                thread::sleep(Duration::from_millis(100));
+                continue;
+            }
+        }
+        return exec_err!("{}", e);
+    }
+    Ok(())
+}
+
+fn create_writing_thread(
+    file_path: PathBuf,
+    maybe_header: Option<String>,
+    lines: Vec<String>,
+    waiting_lock: Arc<AtomicBool>,
+    wait_until: usize,
+) -> JoinHandle<()> {
+    // Timeout for a long period of BrokenPipe error
+    let broken_pipe_timeout = Duration::from_secs(10);
+    let sa = file_path.clone();
+    // Spawn a new thread to write to the FIFO file
+    #[allow(clippy::disallowed_methods)] // spawn allowed only in tests
+    spawn_blocking(move || {
+        let file = OpenOptions::new().write(true).open(sa).unwrap();
+        // Reference time to use when deciding to fail the test
+        let execution_start = Instant::now();
+        if let Some(header) = maybe_header {
+            write_to_fifo(&file, &header, execution_start, broken_pipe_timeout).unwrap();
+        }
+        for (cnt, line) in lines.iter().enumerate() {
+            while waiting_lock.load(Ordering::SeqCst) && cnt > wait_until {
+                thread::sleep(Duration::from_millis(50));
+            }
+            write_to_fifo(&file, line, execution_start, broken_pipe_timeout).unwrap();
+        }
+        drop(file);
+    })
+}
+
+/// This example demonstrates a scanning against an Arrow data source (JSON) and
+/// fetching results
+#[tokio::main]
+async fn main() -> Result<()> {
+    // Create session context
+    let config = SessionConfig::new()
+        .with_batch_size(TEST_BATCH_SIZE)
+        .with_collect_statistics(false)
+        .with_target_partitions(1);
+    let ctx = SessionContext::new_with_config(config);
+    let tmp_dir = TempDir::new()?;
+    let fifo_path = create_fifo_file(&tmp_dir, "fifo_unbounded.csv")?;
+
+    let mut tasks: Vec<JoinHandle<()>> = vec![];
```

Comment:
BTW using a `JoinSet` https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html might be easier here (it encapsulates managing tasks like this)

Reply:
Yes, I was actually planning on making this change. Thanks for reminding me